bokeh/io/notebook.py

```diff
@@ -454,17 +457,23 @@ def show_app(app, state, notebook_url, port=0):
 
     loop = IOLoop.current()
 
+    if not proxy_url_func:
+        proxy_url_func = _server_url
+
```

Comment:
Instead of this check, couldn't `_server_url` just be the default arg value (instead of `None`)?

Reply:
Excellent point, but the definition of the function is above that of _server_url.  It seems like in the code you generally want to keep the private functions at the bottom, but if the definition of _server_url was moved above, I can do that.  I've had notoriously bad luck passing complex default parameters in python (like the dreaded default list argument)

Reply:
I would actually suggest this:

* un-private it, i.e. rename to `default_server_url` and mark `@dev`
* make it a required parameter to `show_app`
* import `default_server_url` in `bokeh.io.showing` and make it the default arg to `show`

I also think `proxy_url_func` should be renamed to `server_url_function` since there may be cases when this hook is useful that do not involve proxies.

Reply:
Sounds good.  Naming is one of the hardest things in CS, so typically I let other people come up with the names.

Reply:
Actually, would it be feasible to let `notebook_url` accept a callable in addition to a string? In which event the callable is responsible for generating the right URL. If this is possible, I think I'd prefer it to adding yah another parameter to `show` that only applies in a subset of use scenarios.

Reply:
i.e., `show(app, notebook_url=my_url_function)`  If a string is passed, the current code path with `_server_url` is used, if a function is passed, it is delegated to entirely, and `_server_url` is not called.

Reply:
Sure that seems reasonable, I think it's mostly up to style (some people really don't like overloading their parameters too much).

Reply:
If JupyterHub follows a common format, we could provide a `jupyter_hub` convenience wrapper that returns a url function that is properly configured:
```
show(app, notebook_url=jupyter_hub("https://foo.com"))
```

Reply:
OK, let's see what that looks like. I definitely agree that in most cases overloading is to be avoided, but I think there is a balance here to not add more parameters to `show`

Reply:
Totally up to you.

As for providing the function, you can see what my definition is here:
https://github.com/lsst-epo/hr-diagram-investigations/blob/0148deaefe038dd3d43b5e6c980d8a5fee4661ec/notebooks/hr/__init__.py#L24

I'm not sure how many of those environment variables are automatically put there by jupyter (I believe this might be up to configuration).  Also it requires installing nbserverproxy (a notebook extension).

I'm happy to provide my proxy function though, it wouldn't even need to take a parameter, which is good, because partial functions make me crazy.  The reason why it has to be a function rather than a static URL is because the port number is going to end up in the URL and isn't known when calling show if port=0.  You want to pick port=0 though, so you don't have to worry about the port conflicts / being already bound from a previous run of the cell.  This means that you have to generate the URL after binding to the port, hence the function call.

Reply:
OK if it's not known to be fully general (i.e. because env vars may or may not be there) then we might instead add docs/examples that demonstrate your jupyterhub function, instead of adding it directly into the core library.

Reply:
Jupyter sets JUPYTERHUB_SERVICE_PREFIX, but EXTERNAL_URL is something that I'm setting in my OAuth spawner configuration (and which @cbanek is inheriting from me):

`pod_env['EXTERNAL_URL'] = oauth_callback[:-len(endstr)]`